blogs/qserial.md

```diff
@@ -0,0 +1,243 @@
+<pre>
+$q.when()
+.then(function () { 
+  console.log("foo"); 
+  var bazDeferred = $q.defer();
+  var bazPromise = bazDeferred.promise;
+  return bazPromise;
+})
+.then(function () { console.log("bar"); });
+</pre>
+<pre>
+foo  
+</pre>
+<br/>
+As expected, our first example logs "foo" then "bar".  However, our second example only logs "foo".  $q does not continue executing the chain of promises until bazPromise has been resolved.  So "bar" is never logged.
+
+This is good!  $q allows us to execute chained promises sequentially by returning promises in our success functions.
+
+Let's see this $q feature in action:
+<pre>
+$q.when()
+.then(function () {
+  var deferred = $q.defer();
+  console.log("foo"); 
+  $timeout(function () { deferred.resolve(); }, 5000);
+  return deferred.promise;
+})
+.then(function () { console.log("bar"); });
+</pre>
+<pre>
+foo  
+//5 seconds
+bar
+</pre>
+<br/>
+Great, so now we see how we can execute promises sequentially.  How do we pass data through to the next promise?  Easy, we pass arguments to .resolve() or .reject().
+
+<pre>
+$q.when()
+.then(function () {
+  var deferred = $q.defer();
+  console.log("foo"); 
+  $timeout(function () { deferred.resolve("bar"); }, 5000);
+  return deferred.promise;
+})
+.then(function (data) { console.log(data); })
+</pre>
+<pre>
+foo  
+//5 seconds
+bar
+</pre>
+<br/>
+Ok, so we've only chained two sequential promises together.  What if I want to chain 5, 10, or 100?  Can we generalize this construct?  A little harder, but sure we can!
+
+<pre>
+function serial(tasks) {
+  var prevPromise;
+  angular.forEach(tasks, function (task) {
+    //First task
+    if (!prevPromise) { 
+      prevPromise = task(); 
+    } else {
+      prevPromise = prevPromise.then(task); 
+    }
+  });
+  return prevTask;
+}
+
+var task1 = function () {
+  var task1Deferred = $q.defer();
+  $timeout(function () {
+    task1Deferred.resolve("foo");
+  }, 5000);
+  return task1Deferred.promise;
+};
+
+var task2 = function (data) { console.log(data); };
+
+serial([task1, task2]);
+</pre>  
+
+<pre>
+foo
+//after 5 seconds
+bar  
+</pre>  
+<br/>
+Great, task1 is executed.  Then task2 is executed after task1Promise is resolved.  
+
+Hold up, I just used a new word "task."  IMO we in the AngularJS community don't use the word "task" as often as we should.  So let me define how I imagine "task."  In plain english a task is a unit of work to be done.  With reference to AngularJS, I often of a task as a function that returns a promise.
+
+Why? Because a function performs a unit of work (or at least kicks off a unit of work).  And a promise let's me know the status of the task (in progress, succeeded, failed).  Knowing the status of a task is very useful!  
+
+The term isn't perfect, but it helps me organize my thoughts when thinking about promises.  So let's stick with it.
+
+<pre>
+function washDishesTask() {
+  var deferred = $q.defer();
+  $timeout(function () { 
+    washDishes(); 
+    $q.defer(); 
```

Comment:
This was meant to be $q.resolve.  But again $timeout's promise is pretty useless in a lot of cases...especially when you're hitting the server in a timeout.